server/src/main/java/io/crate/expression/tablefunctions/MatchesFunction.java

```diff
@@ -0,0 +1,236 @@
+        String value = (String) args[0].value();
+        String pattern = (String) args[1].value();
+        if (value == null || pattern == null) {
+            return () -> NO_RESULTS_ITERATOR;
+        }
+
+        RegexMatcher matcher;
+        if (regexMatcher == null) {
+            String flags = null;
+            if (args.length == 3) {
+                flags = (String) args[2].value();
+            }
+            matcher = createMatcher(pattern, flags);
+        } else {
+            matcher = regexMatcher;
+        }
+
+        Iterator<Row> rowIterator;
+        List<List<String>> rowGroups = matcher.match(value);
+        if (rowGroups == null) {
+            rowIterator = NO_RESULTS_ITERATOR;
+        } else {
+            rowIterator = new Iterator<>() {
+
+                final Object [] columns = new Object[]{ null };
+                final RowN row = new RowN(columns);
+                int idx = 0;
+
+                @Override
+                public boolean hasNext() {
+                    return idx < rowGroups.size();
+                }
+
+                @Override
+                public Row next() {
+                    if (!hasNext()) {
+                        throw new NoSuchElementException("no more rows");
+                    }
+                    columns[0] = rowGroups.get(idx++);
+                    return row;
+                }
+            };
+        }
+        return () -> rowIterator;
+    }
+
+    private static RegexMatcher createMatcher(String regex, @Nullable String flags) {
+
+        Function<Supplier<List<String>>, List<List<String>>> collector =
+            RegexMatcher.isGlobal(flags) ?
+                (supplier) -> {
+                    List<List<String>> result = Stream
+                        .generate(supplier)
+                        .takeWhile(Objects::nonNull)
+                        .collect(Collectors.toList());
+                    return result.isEmpty() ? null : result;
+                }
+                :
+                (supplier) -> {
+                    List<String> groups = supplier.get();
+                    return groups != null ? List.of(groups) : null;
+                };
+        return new RegexMatcher(regex, flags) {
+
+            @Override
+            public List<List<String>> match(String term) {
+                Matcher matcher = initMatcher(term);
+                return collector.apply(() -> {
+                    if (!matcher.find()) {
+                        return null;
+                    }
+                    int groupCount = matcher.groupCount();
+                    if (groupCount == 0) {
+                        try {
+                            return List.of(matcher.group());
+                        } catch (IllegalStateException e) {
+                            return null;
+                        }
+                    }
+                    return IntStream
```

Comment:
imho simple for loop would be more readable and it would avoid to/from object operations (not sure whether it is the case)